Objects/codeobject.c

```diff
@@ -134,105 +134,190 @@ should_intern_string(PyObject *o)
 static PyObject *intern_one_constant(PyObject *op);
 #endif
 
-static int
-intern_strings(PyObject *tuple)
+/* Helper for intern_names and intern_constants, which create a new tuple
+ * as a copy of `orig_tuple`, with some elements possibly changed.
+ * But we optimize for cases where no change, and no copying, is necessary.
+ *
+ * This helper sets the i'th element to `new_item`.
+ * `new_item` is stolen.
+ * Return 0 on success; -1 with exception set on error.
+ *
+ * Error pass-through: If `new_item` is NULL, an exception should be set;
+ * return -1 with that exception still set.
+ *
+ * Until a copy is necessary, `*new_tuple_p` should be NULL.
+ * This function will initialize it to a new reference when needed.
+ *
+ * If `*new_tuple_p` remains NULL after all calls to _set_newtuple_item,
+ * the final result should be a new reference to `orig_tuple`.
+ */
+static inline int
+_set_newtuple_item(
+    PyObject *orig_tuple,
+    PyObject **new_tuple_p /* NULL or owned by caller */,
+    Py_ssize_t i,
+    PyObject *new_item /* Stolen. If NULL, pass error through. */)
+{
+    if (new_item == NULL) {
+        return -1;
+    }
+    if (new_item == PyTuple_GET_ITEM(orig_tuple, i)) {
+        Py_DECREF(new_item);
+        return 0;
+    }
+    PyObject *new_tuple = *new_tuple_p;
+    if (!new_tuple) {
+        Py_ssize_t size = PyTuple_GET_SIZE(orig_tuple);
+        new_tuple = PyTuple_New(size);
+        if (!new_tuple) {
+            Py_DECREF(new_item);
+            return -1;
+        }
+        for (Py_ssize_t i = size; --i >= 0; ) {
+            _PyTuple_ITEMS(new_tuple)[i] = Py_NewRef(
+                _PyTuple_ITEMS(orig_tuple)[i]);
+        }
+        *new_tuple_p = new_tuple;
```

Comment:
Copying the tuple should be a rare event, mostly for creating code from user data.
Should it use some kind of faster-cpython stat counter?